julia/bin/upchan.sh

```diff
@@ -0,0 +1,37 @@
+#!/bin/bash
+
+set -euxo pipefail
+
+scriptdir=$(dirname "$0")
+cd "$scriptdir/.."
+
+Ufactors='16 32 64 128'
+
+# Delete previous output (so that we don't accidentally re-use it)
+for U in $Ufactors; do
+    rm -f ../lib/cuda/cudaUpchannelizer_U$U.cpp
+    rm -f ../lib/cuda/kernels/Upchannelizer_U$U.jl
+    rm -f ../lib/cuda/kernels/Upchannelizer_U$U.ptx
+    rm -f ../lib/cuda/kernels/Upchannelizer_U$U.yaml
+done
+
+# Generate kernel
+for U in $Ufactors; do
+    julia --project=@. --optimize kernels/upchan-U$U.jl 2>&1 | tee output-A40/upchan-U$U.out
+done
+
+# Format generated C++ code
+for U in $Ufactors; do
+    clang-format -i output-A40/upchan-U$U.cxx
+done
+
+# Format generated Julia code
+julia --project=@. --eval 'using JuliaFormatter; JuliaFormatter.format_file("output-A40")'
+
+# Copy kernel into Kotekan
+for U in $Ufactors; do
+    cp output-A40/upchan-U$U.cxx ../lib/cuda/cudaUpchannelizer_U$U.cpp
```

Comment:
There are a fair number of assumptions about this using an A40.  Is the logic specific enough to the A40 that basically that's the only card this will run on, or would it be possible to be more general for say an A10/A30/L40/L40S, etc. if those endup getting used for some instrument?

Reply:
I don't know, but these kernels rely heavily on the exact set (and performance levels) of tensor core instructions, so I'm guessing that if a different card got used, one would have to carefully check the performance anyway, so that there's not a ton of value in theoretically targeting a broader set of cards.